# Preparation and transformation of the data

> In this initial section, before diving into our main analysis, it's crucial to properly prepare and understand our dataset that form the foundation of our study. This preliminary phase involves several key steps:
> 1. Data Loading: We'll begin by opening our data files and adding headers based on the *Readme* of thE *CMU Movie Summary Corpus* study (http://www.cs.cmu.edu/~ark/personas/).
> 2. Data Size Evaluation: Understanding the size of our datasets is essential for determining appropriate processing methods and potential computational constraints.
> 3. Data Merging: We'll combine two relevant dataframes (`characters_df` and `movies_df`) to create a more comprehensive dataset and to facilitate the statistical anaylsis.
> 4. Separation of data into decades: After merging, we'll separate the dataset into decades as we want to study an evolution over time.

In [8]:
import pandas as pd
import numpy as np

In [18]:
# Script to load the notebook utils.ipynb
import nbformat
from IPython.core.interactiveshell import InteractiveShell

# Load the notebook utils.ipynb
with open('../scripts/utils.ipynb') as f:
    nb = nbformat.read(f, as_version=4)

# Create an instance of InteractiveShell
shell = InteractiveShell.instance()

# Execute the notebook utils.ipynb
for cell in nb.cells:
    if cell.cell_type == 'code':
        shell.run_cell(cell.source)

## 1. Adding dataset headers

In [19]:
import pandas as pd

# Characters file
characters_df = pd.read_csv('../../data/character.metadata.tsv', sep='\t', header=None)
characters_df.columns = ['Wikipedia_movie_ID', 'Freebase_Movie_ID', 'Movie_release_date', 'Character_name', 'Actor_date_of_birth', 'Actor_gender', 'Actor_height','Actor_ethnicity',
                         'Actor_name', 'Actor_age_at_movie_release','Freebase_character/actor_map_ID','Freebase_character_ID','Freebase_actor_ID']
characters_df.head()
characters_df.to_csv('character.metadata.tsv',  sep='\t', index=False)

# Movies file
movies_df = pd.read_csv('../../data/movie.metadata.tsv', sep='\t', header=None)
movies_df.columns = ['Wikipedia_movie_ID', 'Freebase_Movie_ID', 'Movie_name', 'Movie_release_date', 'Movie_box_office_revenue', 'Movie_runtime', 'Movie_languages','Movie_countries',
                         'Movie_genres']
movies_df.to_csv('movie.metadata.tsv',  sep='\t', index=False)

# Name clusters file
path_name_cluster = '../../data/name.clusters.txt'
name_cluster_df = pd.read_csv(path_name_cluster, delimiter='\t', header=None)
name_cluster_df.columns = ['unique_character_name', 'freebase_actor_id']
name_cluster_df.to_csv('name.clusters.txt',  sep='\t', index=False)

# Plot summary file
# Read the txt file by specifying the delimiter (here, a tabulation)
path_plot = '../../data/plot_summaries.txt'
plot_summary_df = pd.read_csv(path_plot, delimiter='\t', header=None)
plot_summary_df.columns = ['movie_id', 'plot_summary']
plot_summary_df.to_csv('plot_summaries.txt',  sep='\t', index=False)

# TV tropes cluster
path_name_cluster = '../../data/tvtropes.clusters.txt'
tv_tropes_df = pd.read_csv(path_name_cluster, delimiter='\t', header=None)
tv_tropes_df.columns = ['character_types', 'ID_field']
tv_tropes_df.to_csv('tvtropes.clusters.txt',  sep='\t', index=False)

# 2. Merging of the `characters_df` and `moves_df`

In [20]:
merged_df = pd.merge(characters_df, movies_df, on=['Wikipedia_movie_ID', 'Freebase_Movie_ID', 'Movie_release_date'])
#merged_df.to_csv('merged_df.tsv', sep='\t', index=False)

In [21]:
merged_df.head()    

Wikipedia_movie_ID Freebase_Movie_ID Movie_release_date  \
0              975900         /m/03vyhn         2001-08-24   
1              975900         /m/03vyhn         2001-08-24   
2              975900         /m/03vyhn         2001-08-24   
3              975900         /m/03vyhn         2001-08-24   
4              975900         /m/03vyhn         2001-08-24   

               Character_name Actor_date_of_birth Actor_gender  Actor_height  \
0                    Akooshay          1958-08-26            F         1.620   
1  Lieutenant Melanie Ballard          1974-08-15            F         1.780   
2         Desolation Williams          1969-06-15            M         1.727   
3          Sgt Jericho Butler          1967-09-12            M         1.750   
4             Bashira Kincaid          1977-09-25            F         1.650   

  Actor_ethnicity          Actor_name  Actor_age_at_movie_release  \
0             NaN      Wanda De Jesus                        42.0   
1      /m/044038p  Natasha Henstridge                        27.0   
2         /m/0x67            Ice Cube                        32.0   
3             NaN       Jason Statham                        33.0   
4             NaN         Clea DuVall                        23.0   

  Freebase_character/actor_map_ID Freebase_character_ID Freebase_actor_ID  \
0                      /m/0bgchxw            /m/0bgcj3x        /m/03wcfv7   
1                       /m/0jys3m            /m/0bgchn4         /m/0346l4   
2                       /m/0jys3g            /m/0bgchn_        /m/01vw26l   
3                      /m/02vchl6            /m/0bgchnq         /m/034hyc   
4                      /m/02vbb3r            /m/0bgchp9         /m/01y9xg   

       Movie_name  Movie_box_office_revenue  Movie_runtime  \
0  Ghosts of Mars                14010832.0           98.0   
1  Ghosts of Mars                14010832.0           98.0   
2  Ghosts of Mars                14010832.0           98.0   
3  Ghosts of Mars                14010832.0           98.0   
4  Ghosts of Mars                14010832.0           98.0   

                      Movie_languages  \
0  {"/m/02h40lc": "English Language"}   
1  {"/m/02h40lc": "English Language"}   
2  {"/m/02h40lc": "English Language"}   
3  {"/m/02h40lc": "English Language"}   
4  {"/m/02h40lc": "English Language"}   

                             Movie_countries  \
0  {"/m/09c7w0": "United States of America"}   
1  {"/m/09c7w0": "United States of America"}   
2  {"/m/09c7w0": "United States of America"}   
3  {"/m/09c7w0": "United States of America"}   
4  {"/m/09c7w0": "United States of America"}   

                                        Movie_genres  
0  {"/m/01jfsb": "Thriller", "/m/06n90": "Science...  
1  {"/m/01jfsb": "Thriller", "/m/06n90": "Science...  
2  {"/m/01jfsb": "Thriller", "/m/06n90": "Science...  
3  {"/m/01jfsb": "Thriller", "/m/06n90": "Science...  
4  {"/m/01jfsb": "Thriller", "/m/06n90": "Science...

# 3. Separation of the `merged_df` into decades

In [22]:
# Conversion of the 'Movie_release_date' column to datetime format
merged_df['Movie_release_date'] = pd.to_datetime(merged_df['Movie_release_date'], errors='coerce')

# Count the number of films without release date
films_without_release_date = merged_df['Movie_release_date'].isna().sum()

# Display the number of films without release date
print(f"Number of movies without release date : {films_without_release_date}, which is {round(films_without_release_date/len(merged_df)*100, 2)}% of the dataset")

# Count the number of films without associated genres
films_without_genre = merged_df['Movie_genres'].isna().sum()

# Display the number of films without associated genres
print(f"Number of movies without genre : {films_without_genre}, which is {round(films_without_genre/len(merged_df)*100, 2)}% of the dataset")

Number of movies without release date : 176797, which is 39.23% of the dataset
Number of movies without genre : 0, which is 0.0% of the dataset


> As 39.23% of the dataset has no release date, we cannot simply discard it. We decide to use the average of the release years of films in the same genre, given that all the films in the dataset have a genre or a combination of genres. For the combination of genres, we take the average of the averages for each genre.

In [23]:
# Creation of a new column called "genre_list" in movies_df to have a list of all genres of a movie
movies_df['genre_list'] = movies_df['Movie_genres'].astype(str).apply(extract_string)

# Creation of a new column called "genre_list" in merged_df to have a list of all genres of a movie
movies_df['genre_list'] = movies_df['Movie_genres'].astype(str).apply(extract_string)
merged_df['genre_list'] = merged_df['Movie_genres'].astype(str).apply(extract_string)

# Conversion of dates to years (to unfiformize the format)
merged_df['Movie_release_year'] = merged_df['Movie_release_date'].dt.year

# Calculate the mean release year by genre
mean_release_year_by_genre = merged_df.explode('genre_list').groupby('genre_list')['Movie_release_year'].mean()

# Estimation of the release year of the films without release date
merged_df['Estimated_release_year'] = merged_df.apply(estimate_release_year, axis=1, args=(mean_release_year_by_genre,))

# Vérification
films_without_release_date = merged_df['Estimated_release_year'].isna().sum()
print(f"Number of movies without release date : {films_without_release_date} movies, which is {round(films_without_release_date/len(merged_df)*100, 2)}% of the dataset")

Number of movies without release date : 2879 movies, which is 0.64% of the dataset


In [24]:
merged_df.to_csv('merged_df.tsv', sep='\t', index=False)
mean_release_year_by_genre.to_csv('mean_release_year_by_genre.tsv', sep='\t', index=False)

> After using the mean release date of movies with similar genres or the mean of the means for genres combination, only 0.64% of the dataset, which corresponds to 2879 movies, still don't have release date. This is acceptable to work with it. What could explain why some movies still don't have release date is due to the used method. Inded, the mean meathod is not applicable on movies that have a unique genre and no release date, making it impossible to calculate the average.

# Data treatment of cluster correlations

> This code processes correlations derived from the SBERT and LLAMA clustering workflow. It extracts the highest correlations from the data, which are then used to generate plots.

In [15]:
# Create a CSV file with correlations > 0.35
higher_correlations = []
for decade in np.arange(1900, 2010, 10):

    df = pd.read_csv(f'heatmaps_data/heatmap_data_{decade}.csv', index_col=0)

    # Convert all columns to numeric, forcing errors to NaN 
    df = df.apply(pd.to_numeric, errors='coerce')

    # Use stack() to flatten the dataframe and filter correlations greater than 0.35
    filtered_df = df.stack()
    filtered_df = filtered_df[filtered_df > 0.35]

    # Loop through the values and add them to the higher_accuracy list
    for (row_label, col_label), correlation_value in filtered_df.items():
        higher_correlations.append({
            'Decade': decade,
            'Movies cluster': row_label,
            'History cluster': col_label,
            'Correlation': correlation_value
        })

# Create a dataframe from the higher_accuracy list
higher_correlations_df = pd.DataFrame(higher_correlations)

# Save the result on csv
higher_correlations_df.to_csv('higher_correlations_data.csv', index=False)

In [16]:
# Find the max similarities for each decade
max_similarities = []

for decade in np.arange(1900, 2010, 10):
   
    df = pd.read_csv(f'max_similarity_plots_data/bar_plot_data_{decade}.csv', index_col=0)

    df = df.apply(pd.to_numeric, errors='coerce')

    # Find the maximum value in the dataframe
    max_value = df.max().max()

    # Row and column labels for the maximum value
    row_label, col_label = np.unravel_index(df.values.argmax(), df.shape)

    # Add the result to the list
    max_similarities.append({
        'Decade': decade,
        'History cluster': df.index[row_label],
        'Similarity': max_value
    })

# Create a dataframe from the max_similarities list
max_similarities_df = pd.DataFrame(max_similarities)

# Save the result 
max_similarities_df.to_csv('max_similarities_data.csv', index=False)